Source/Core/Core/IOS/USB/Bluetooth/LibUSBBluetoothAdapter.cpp

```diff
@@ -0,0 +1,570 @@
+
+#include "Core/IOS/USB/Bluetooth/LibUSBBluetoothAdapter.h"
+
+#include <algorithm>
+#include <cstdlib>
+#include <cstring>
+#include <ranges>
+#include <utility>
+
+#include <fmt/format.h>
+#include <libusb.h>
+
+#include "Common/MsgHandler.h"
+
+#include "Core/Config/MainSettings.h"
+#include "Core/Core.h"
+#include "Core/IOS/USB/Bluetooth/hci.h"
+#include "Core/IOS/USB/Host.h"
+
+namespace
+{
+constexpr std::size_t BUFFER_SIZE = 1024;
+
+constexpr auto HCI_COMMAND_TIMEOUT = std::chrono::milliseconds{1000};
+
+constexpr u8 REQUEST_TYPE =
+    u8(u8(LIBUSB_ENDPOINT_OUT) | u8(LIBUSB_REQUEST_TYPE_CLASS)) | u8(LIBUSB_RECIPIENT_DEVICE);
+
+constexpr u8 HCI_EVENT = 0x81;
+constexpr u8 ACL_DATA_IN = 0x82;
+
+template <auto MemFun>
+constexpr libusb_transfer_cb_fn LibUSBMemFunCallback()
+{
+  return [](libusb_transfer* tr) {
+    std::invoke(MemFun, static_cast<Common::ObjectType<MemFun>*>(tr->user_data), tr);
+  };
+}
+
+bool IsBluetoothDevice(const libusb_device_descriptor& descriptor)
+{
+  constexpr u8 SUBCLASS = 0x01;
+  constexpr u8 PROTOCOL_BLUETOOTH = 0x01;
+  // Some devices misreport their class, so we avoid relying solely on descriptor checks and allow
+  // users to specify their own VID/PID.
+  return LibUSBBluetoothAdapter::IsConfiguredBluetoothDevice(descriptor.idVendor,
+                                                             descriptor.idProduct) ||
+         (descriptor.bDeviceClass == LIBUSB_CLASS_WIRELESS &&
+          descriptor.bDeviceSubClass == SUBCLASS &&
+          descriptor.bDeviceProtocol == PROTOCOL_BLUETOOTH);
+}
+
+}  // namespace
+
+bool LibUSBBluetoothAdapter::IsWiiBTModule() const
+{
+  return m_is_wii_bt_module;
+}
+
+bool LibUSBBluetoothAdapter::IsConfiguredBluetoothDevice(u16 vid, u16 pid)
+{
+  const int configured_vid = Config::Get(Config::MAIN_BLUETOOTH_PASSTHROUGH_VID);
+  const int configured_pid = Config::Get(Config::MAIN_BLUETOOTH_PASSTHROUGH_PID);
+  return configured_vid == vid && configured_pid == pid;
+}
+
+LibUSBBluetoothAdapter::LibUSBBluetoothAdapter()
+{
+  if (!m_context.IsValid())
+    return;
+
+  m_last_open_error.clear();
+  const int ret = m_context.GetDeviceList([this](libusb_device* device) {
+    libusb_device_descriptor device_descriptor;
+    libusb_get_device_descriptor(device, &device_descriptor);
+    auto [make_config_descriptor_ret, config_descriptor] =
+        LibusbUtils::MakeConfigDescriptor(device);
+    if (make_config_descriptor_ret != LIBUSB_SUCCESS || !config_descriptor)
+    {
+      ERROR_LOG_FMT(IOS_WIIMOTE, "Failed to get config descriptor for device {:04x}:{:04x}: {}",
+                    device_descriptor.idVendor, device_descriptor.idProduct,
+                    LibusbUtils::ErrorWrap(make_config_descriptor_ret));
+      return true;
+    }
+
+    if (IsBluetoothDevice(device_descriptor) && OpenDevice(device_descriptor, device))
+    {
+      unsigned char manufacturer[50] = {}, product[50] = {}, serial_number[50] = {};
```

Comment:
Any significance to the number 50 here?

Reply:
Probably not. It's moved from existing code that's been like that since the original Bluetooth Passthrough commit. It's ultimately just used for a log entry.